mava/wrappers/gigastep.py

```diff
@@ -0,0 +1,304 @@
+from jumanji.wrappers import Wrapper
+
+from mava.types import Observation, ObservationGlobalState, State
+
+if TYPE_CHECKING:  # https://github.com/python/mypy/issues/6239
+    from dataclasses import dataclass
+else:
+    from flax.struct import dataclass
+
+
+@dataclass
+class GigastepState:
+    """Wrapper around a Gigastep state to provide necessary attributes for Jumanji environments."""
+
+    state: State
+    key: PRNGKey
+    step: int
+    adversary_action: Array
+
+
+class GigastepWrapper(Wrapper):
+    def __init__(
+        self,
+        env: GigastepEnv,
+        time_limit: int = 500,
+        has_global_state: bool = False,
+    ):
+        """
+        Wraps a Gigastep environment so that its API is compatible with Jumanji environments.
+
+        Args:
+            env: The Gigastep environment to be wrapped.
+            time_limit (int): The maximum duration of each episode, in seconds. Defaults to 500.
+            has_global_state (bool): Whether the environment has a global state. Defaults to False.
+        """
+
+        super().__init__(env)
+        assert (
+            env.discrete_actions
+        ), "Only discrete action spaces are currently supported for Gigastep environments"
+        assert (
+            env._obs_type == "vector"
+        ), "Only Vector observations are currently supported for Gigastep environments"
+
+        self._env: GigastepEnv
+        self._env.max_episode_length = time_limit
+        self.num_agents = self._env.n_agents_team1
+        self.num_actions = self._env.n_actions
+
+        self.has_global_state = has_global_state
+
+    def reset(self, key: PRNGKey) -> Tuple[GigastepState, TimeStep]:
+        """
+        Reset the Gigastep environment.
+
+        Args:
+            key (PRNGKey): The PRNGKey.
+
+        Returns:
+            GigastepState : the state of the environment.
+            TimeStep : the first time step.
+
+        """
+
+        key, reset_key, adversary_key = jax.random.split(key, 3)
+        obs, state = self._env.reset(reset_key)
+
+        team1_obs, team1_state, team2_obs, team2_state = self._split_obs_and_state(obs, state)
+
+        adversary_actions = self.adversary_policy(team2_obs, team2_state, adversary_key)
+
+        if self.has_global_state:
+            obs = ObservationGlobalState(
+                agents_view=team1_obs,
+                action_mask=self.action_mask(),
+                global_state=self.get_global_state(obs, adversary_actions),
+                step_count=jnp.zeros(self.num_agents, dtype=int),
+            )
+        else:
+            obs = Observation(
+                agents_view=team1_obs,
+                action_mask=self.action_mask(),
+                step_count=jnp.zeros(self.num_agents, dtype=int),
+            )
+
+        state = GigastepState(state, key, 0, adversary_actions)
+        timestep = restart(obs, shape=(self.num_agents,), extras={"won_episode": False})
+        return state, timestep
```

Comment:
Can you please refer to the logic made here to reduce redundant code in [jaxmarl wrapper](https://github.com/instadeepai/Mava/blob/f23fa5890ad8036aa8bc844a412c6997aa519c35/mava/wrappers/jaxmarl.py#L232).